include/zephyr/bluetooth/encrypted_ad_data.h

```diff
@@ -0,0 +1,77 @@
+/* Copyright (c) 2023 Nordic Semiconductor ASA
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+#include <stddef.h>
+#include <stdint.h>
+
+#include <zephyr/kernel.h>
+
+#define BT_EAD_AAD_SIZE	       1  /**< Additional Authenticated Data size in bytes */
+#define BT_EAD_RANDOMIZER_SIZE 5  /**< Randomizer size in bytes */
+#define BT_EAD_NONCE_SIZE      13 /**< nonce size in bytes */
+#define BT_EAD_KEY_SIZE	       16 /**< Key size in bytes */
+#define BT_EAD_IV_SIZE	       8  /**< Initialisation Vector size in bytes */
+#define BT_EAD_MIC_SIZE	       4  /**< MIC size in bytes */
+
+/* This value is used to set the directionBit of the CCM  nonce to the MSB of the Randomizer field
+ * (see Supplement to the Bluetooth Core   Specification v11, Part A 1.23.3)
+ */
+#define BT_EAD_RANDOMIZER_DIRECTION_BIT 7
+
+/* Fixed value used for the Additional Authenticated Data (see Supplement to the Bluetooth Core
+ * Specification v11, Part A 1.23.3)
+ */
+static uint8_t bt_ead_aad[] = {0xEA};
+BUILD_ASSERT(sizeof(bt_ead_aad) == BT_EAD_AAD_SIZE);
+
+/**
+ * @brief Encrypt and authenticate the given advertising data.
+ * The resulting data in \p encrypted_ad_data will look like that:
+ *     - Randomizer is added in the 5 first bytes;
+ *     - AAD is added in the next byte;
+ *     - Encrypted payload is added (payload_size bytes);
+ *     - MIC is added in the last 4 bytes.
+ *
+ * @note The function must be called each time the RPA is updated or the data
+ * are modified.
+ *
+ * @note The term `advertising strucure` is used to describe the advertising
+ *       data with the advertising type and the length of those two.
+ *
+ * @param[in]  session_key Key of 16 bytes used for the encryption.
+ * @param[in]  iv Initialisation Vector used to generate the nonce. It must be
+ *             changed each time the Session Key change.
+ * @param[in]  payload Advertising Data to encrypt. Can be multiple advertising
+ *             structure concatenated.
+ * @param[in]  payload_size Size of the Advertising Data to encrypt.
+ * @param[out] encrypted_ad_data Encrypted Adv Data including the Randomizer and
+ *             the MIC. Size must be `BT_EAD_RANDOMIZER_SIZE + BT_EAD_AAD_SIZE +
+ *             payload_size + BT_EAD_MIC_SIZE`.
+ * @return int Size of \p encrypted_ad_data in bytes if successful, else return
+ *             `-EIO` if an error occurred during the encryption.
+ */
+int bt_ead_encrypt_ad_data(const uint8_t *session_key, const uint8_t *iv, const uint8_t *payload,
```

Comment:
For constant-size arrays we've often used the size of the arrays in the declaration like
```suggestion
int bt_ead_encrypt_ad_data(const uint8_t session_key[16], const uint8_t iv[BT_EAD_IV_SIZE], const uint8_t *payload,
```

Reply:
Another point about the API name: both "ad" and "data" are already part of the "ead" acronym itself, so they're redundant. The appropriate name would therefore be simply `bt_ead_encrypt()`